### PR TITLE
Issue 1182/editable title type

### LINKED ISSUE #1182: Editable event title and type 🤘
## Description
A user should be able to pick either a title or a type, or both for an event.

The term "type" refers to what is sometimes referred to in the UI as "category", and what is technically called "activity" in the Zetkin API, i.e. the piece of information on an event (events being referred to as "action" per Gen2 terminology) that indicate _what_ participants will do at the event.

## Relevant Job Stories
* When I’m creating a new event, I don’t want to forget adding a type, so that later when I use Smart Search or some other indirectly related feature I will find the people who attended the event

## Prerequisites
Don't start this until #1161 is done

## Requirements
- [x] Type should be in the subtitle
  - [x] Editing should work as an `Autocomplete`, with fuzzy search
  - [x] There should be an option called "Uncategorized" which represents `null`
  - [x] When user has typed something, there should be an option called "Create type '<WHAT THE USER TYPED>'"
  - [x] Choosing the "Create" option should create and select the new type
- [x] Title and Type should enter and leave edit mode together – focusing one should take the other into edit mode as well
- [x] Blurring both fields should save
- [x] Editing should adhere to the design specifications below

## Possible implementations
* Maybe create a new Autocomplete version of the `ZUIEditTextInPlace`?
* Retrieve available types using `/orgs/ID/activities`
* Create a new type using `POST /orgs/ID/activities`

### Default state (title and type)
image

### Default state (title but no type)
image

### Default state (type, but no title)
image

### Default state (no title, no type)
image

### Edit mode (title and type)
image

### Edit mode (title, but no type)
image

### Edit mode (type, but no title)
image

### Edit mode (no type, no title)
image

## Open questions
None

### PR DESCRIPTION
**Pull request by kaulfield23**
## Description
This PR implements Autocomplete for event type. When user types texts in autocomplete, it shows option `create "blah blah"` for creating user's type. When one of title or type is on focus, event title (`ZUIEditTextInPlace`) and Autocomplete will be highlighted. If title is empty, it would show event type as title.

## Screenshots

https://user-images.githubusercontent.com/77925373/232531386-6ce22d72-c6af-411d-89d4-5fceb6258f80.mp4

when there is no event type and title
image
image

## Changes

* Adds EventTypeAutocomplete.tsx
* Adds EventTypesModel.ts
* Adds setType in EventDataModel.ts
* Adds showBorder prop to `ZUIEditTextinPlace`
* Adds condition to tooltipText function in `ZUIEditTextInPlace` to show specific tooltip content
* Adds messages to messageIds.

## Notes to reviewer
Finally done...! :grimacing:
It has this minor error with title

https://user-images.githubusercontent.com/77925373/232533791-45d14ff5-351a-4daf-894c-95e49434f3f3.mp4
+ when I select the option, it shortly blinks the former type before showing the selected type. It happened after I had added `blurOnSelect`. If I remove `blurOnSelect` from `autocomplete` then the caret in `autocomplete` will be remained after select the option.

image
Something makes subtitle not visible when I shrink the window.  

## Related issues
Resolves #1182

### CODE REVIEW

File: src/zui/ZUIEditTextInPlace/index.tsx
Comment (richardolsson) [inline src/zui/ZUIEditTextInPlace/index.tsx:55]: `ZUIEditTextInPlace`, being a general-purpose reusable component, should not know anything about _types_, a very specific aspect of the event feature in Zetkin. When you find yourself adding properties like this, named after a very specific feature, the separation of concerns alarm bells should go off.

File: src/zui/ZUIEditTextInPlace/index.tsx
Comment (richardolsson) [inline src/zui/ZUIEditTextInPlace/index.tsx:61]: These look like (and are indeed used like) state from `useState()`, but components should not be designed to directly set other components state. Instead, the common practice is to communicate upwards using `onXYZ` callbacks, that communicate what _happened_, not what the parent component should do in effect. In this case, a name like `onFocus` might make sense, i.e. a callback that communicates to the surroundings that this component was focused and went into edit mode.

File: src/zui/ZUIEditTextInPlace/index.tsx
Comment (richardolsson) [inline src/zui/ZUIEditTextInPlace/index.tsx:34]: There should be no need for this extra class, as there is already styling for the border in the `input` class.

File: src/zui/ZUIEditTextInPlace/index.tsx
Comment (richardolsson) [inline src/zui/ZUIEditTextInPlace/index.tsx:211]: Instead of toggling between different classes, it's possible to use a property to `useStyles()` which allows you to send in parameters to the styling. Here's an example from `ZUIHeader`. 

Style definitions (see lines 24 and 32):
https://github.com/zetkin/app.zetkin.org/blob/e7c6b35082cf3b3d9069033e35453e0ea006f1f3/src/zui/ZUIHeader.tsx#L24-L36

Using the styles:
https://github.com/zetkin/app.zetkin.org/blob/e7c6b35082cf3b3d9069033e35453e0ea006f1f3/src/zui/ZUIHeader.tsx#L76

File: src/zui/ZUIEditTextInPlace/index.tsx
Comment (richardolsson) [inline src/zui/ZUIEditTextInPlace/index.tsx:179]: What is the purpose of all of this logic? To me it seems like it's overkill, and just using `value={editing? text : (text || placeholder)}` seems to work for me, but I might be missing some edge case that you've bumped into?

File: src/zui/ZUIAutocompleteInPlaceTest/index.tsx
Comment (richardolsson) [inline src/zui/ZUIAutocompleteInPlaceTest/index.tsx:109]: I don't think it should be necessary to use `useEditPreviewBlock`, or even `ZUIPreviewableInput` or `ClickAwayListener`. Note how `ZUIEditTextInPlace` don't use either of those, and still is able to do more or less what this component is trying to do.

In my head, using a regular `<Autocomplete>`, styling it differently in different states, and moving between the states `onFocus` and `onBlur` should be enough. But I haven't tried, so maybe it's not possible. I'm thinking we should take a look at it together!

File: src/features/events/layout/EventLayout.tsx
Comment (richardolsson) [inline src/features/events/layout/EventLayout.tsx:35]: What is the purpose of the `autoOnEdit` state, and how is it different from the `showBorder` state?

**Review by richardolsson — COMMENTED** [review]
I have taken the time this morning to take a proper look at your code. This turned out to be quite complex, and you've come really far in solving this issue. But I think it's been made more difficult than it needs to be, and I'm inclined to take a look at it together with you. Let's talk about it!

File: src/features/events/store.ts
Comment (kaulfield23) [reply src/features/events/store.ts:42]: Should I change every reducers name in this file then? like `eventLoad` to `load`, `eventTypesLoad` to `typesLoad`, `eventUpdate` to `update`?

File: src/zui/ZUIAutocompleteInPlaceTest/index.tsx
Comment (kaulfield23) [reply src/zui/ZUIAutocompleteInPlaceTest/index.tsx:1]: Yes I agree, I should have made another branch for this Test one

File: src/zui/ZUIEditTextInPlace/index.tsx
Comment (kaulfield23) [reply src/zui/ZUIEditTextInPlace/index.tsx:201]: It is not something I wrote. I added onClick prop just to stop click away event

File: src/zui/ZUIEditTextInPlace/index.tsx
Comment (kaulfield23) [reply src/zui/ZUIEditTextInPlace/index.tsx:179]: I misunderstood the design, so I thought when user types the event, it would show the text as placeholder in title. So the first screenshot is what I made, and second one is what you suggest.
image

File: src/zui/ZUIAutocompleteInPlaceTest/index.tsx
Comment (kaulfield23) [reply src/zui/ZUIAutocompleteInPlaceTest/index.tsx:109]: Okay!

File: src/features/events/layout/EventLayout.tsx
Comment (kaulfield23) [reply src/features/events/layout/EventLayout.tsx:35]: I found that when user keep clicking title - autocomplete - title autocomplete repeatedly, it would disable border from event title cause `ZUiEditTextInPlace` doesn't know if autocomplete is still in edit mode or not. That was the reason why

**Review by richardolsson — APPROVED** [review]
I think the implementation is messy in some parts, but I know this was a very difficult task and I think we need to move forward. The solution would be to move some of the logic into a model, but I've started rethinking whether models are even helping us or if it would simply be better to use hooks, so I don't think spending the time refactoring it here makes sense until we've made that decision.

Nice work on this!

File: src/utils/types/zetkin.ts
Comment (richardolsson) [reply src/utils/types/zetkin.ts:359]: You don't need to change this right now.

Comment (kaulfield23) [reply src/utils/types/zetkin.ts]: I just renamed it and fixed it after I resolved merge conflict!

**Review by richardolsson — APPROVED** [review]
Nice work!

**Pull request merged**
